### PR TITLE
Add I2C support to nucleo_f411re

### PR DESCRIPTION
**Pull request by rosterloh**
Nucleo board is falling behind so here is a patch to add I2C support

### CODE REVIEW

**Review by dbkinder — CHANGES_REQUESTED** [review]
doc changes look OK, but there are problems with the commit message (see the shippable log https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/5937/3/tests )

File: boards/arm/nucleo_f411re/nucleo_f411re_defconfig [hunk 1]
```diff
@@ -12,6 +12,10 @@ CONFIG_STM32_ARM_MPU_ENABLE=y
 # enable USART2 - passthrough to STLINK v2 connector
 CONFIG_UART_STM32_PORT_2=y
 
+# enable I2C1
+CONFIG_I2C=y
+CONFIG_I2C_STM32_V1=y
+
```
Comment (superna9999) [inline boards/arm/nucleo_f411re/nucleo_f411re_defconfig]: Please also enable I2C2 as documented in peripheral mapping

File: boards/arm/nucleo_f411re/Kconfig.board [hunk 2]
```diff
@@ -7,4 +7,5 @@
 
 config BOARD_NUCLEO_F411RE
 	bool "NUCLEO-64 F411RE Development Board"
+	select I2C_1 if I2C
```
Comment (superna9999) [inline boards/arm/nucleo_f411re/Kconfig.board]: Please remove this line

**Review by superna9999 — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_f411re/Kconfig.board (hunk 2 above)
Comment (rosterloh) [reply boards/arm/nucleo_f411re/Kconfig.board]: This is the same way it is done on the nucleo_f401re and the 96board_carbon. Shouldn't it be standardised in some way?

File: boards/arm/nucleo_f411re/nucleo_f411re_defconfig (hunk 1 above)
Comment (rosterloh) [reply boards/arm/nucleo_f411re/nucleo_f411re_defconfig]: See comment above.

File: boards/arm/nucleo_f411re/Kconfig.board (hunk 2 above)
Comment (superna9999) [reply boards/arm/nucleo_f411re/Kconfig.board]: Concerning the carbon board, it makes sense since the i2c bus is fixed on the board, for nucleo board, the pin selection is dependent on the project you make. So it’s place should be in the defconfig so you can change in the app config without any conflicts whatsoever

File: boards/arm/nucleo_f411re/nucleo_f411re_defconfig (hunk 1 above)
Comment (superna9999) [reply boards/arm/nucleo_f411re/nucleo_f411re_defconfig]: I was referring the i2c you specify in the board documentation, so you should activate it in the defconfig accordingly

Comment (rosterloh) [reply boards/arm/nucleo_f411re/nucleo_f411re_defconfig]: I'm happy to change the board config to not select I2C_1 but to add CONFIG_I2C_1=y and CONFIG_I2C_2=y to the defconfig would be different to every other board. I'm happy to do whatever you want but it seems odd.

**Review by dbkinder — APPROVED** [review]

File: boards/arm/nucleo_f411re/pinmux.c [hunk 3]
```diff
@@ -22,6 +22,14 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PA2, STM32F4_PINMUX_FUNC_PA2_USART2_TX},
 	{STM32_PIN_PA3, STM32F4_PINMUX_FUNC_PA3_USART2_RX},
 #endif	/* CONFIG_UART_STM32_PORT_2 */
+#ifdef CONFIG_I2C_1
+	{STM32_PIN_PB8, STM32F4_PINMUX_FUNC_PB8_I2C1_SCL},
+	{STM32_PIN_PB9, STM32F4_PINMUX_FUNC_PB9_I2C1_SDA},
+#endif /* CONFIG_I2C_1 */
+#ifdef CONFIG_I2C_2
+	{STM32_PIN_PB10, STM32F4_PINMUX_FUNC_PB10_I2C2_SCL},
+	{STM32_PIN_PB3, STM32F4_PINMUX_FUNC_PB3_I2C2_SDA},
```
Comment (erwango) [inline boards/arm/nucleo_f411re/pinmux.c]: Is that PB2 or PB3? 
You mention PB2 in nucleof411re.rst

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_f411re/pinmux.c (hunk 3 above)
Comment (rosterloh) [reply boards/arm/nucleo_f411re/pinmux.c]: You are quite right. Changed to PB3 in nucleof411re.rst

File: boards/arm/nucleo_f411re/nucleo_f411re_defconfig (hunk 1 above)
Comment (superna9999) [reply boards/arm/nucleo_f411re/nucleo_f411re_defconfig]: The idea was to match to rst doc in the defconfig, but not enforce it in the board config

**Review by erwango — COMMENTED** [review]
LGTM appart from comment from superna9999 to be taken into account

**Review by erwango — DISMISSED** [review]

File: boards/arm/nucleo_f411re/Kconfig.board (hunk 2 above)
Comment (erwango) [reply boards/arm/nucleo_f411re/Kconfig.board]: Please take comment into account

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_f411re/Kconfig.board (hunk 2 above)
Comment (erwango) [reply boards/arm/nucleo_f411re/Kconfig.board]: @superna9999: I had a new though about it and I think we can consider providing default pinmuxing for Arduino connectors on Nucleo boards. It would help quick testing with various Arduino compatible shields.
Thought, it would belong to a global new PR taking into account all STM32 boards. So, I agree we need to remove line right now on this PR (to avoid too much divergences amongst boards today)

**Comment by galak** [comment]
@rosterloh any plans to update?

or should I close this out?

**Comment by rosterloh** [comment]
@galak I'm not sure how relevant this is anymore as a lot has changed in zephyr lately but I've rebased the commit

**Review by ydamigos — CHANGES_REQUESTED** [review]
Board configuration should be done as follows:

    in boards/arm/<board>/<board>_defconfig
    -- SoC configuration
    -- Kernel configuration
    -- Driver activation
    -- Specific driver configuration that cannot be done in Kconfig.defconfig due to use of 'choice' type
    in boards/*/<board>/Kconfig.defconfig
    --Driver configuration
    --Configuration of peripherals available directly on board (sensors, communication chip like BT, WiFi, ST Link, ..)

Finally, configuration should be done as follows:

    pin configuration for peripheral available on boards
    pin configuration matching silkscreen, focusing on arduino connectors

**Comment by ydamigos** [comment]
@rosterloh Please also add commit message body and Signed-Off-By.

**Comment by rosterloh** [comment]
@ydamigos I'm not sure I completely understand what you're asking for but I think I've implemented this as per your instructions now

**Review by ydamigos — CHANGES_REQUESTED** [review]
@rosterloh Thanks for the changes. You should also enable I2C_1 in Kconfig.defconfig and I2C in nucleo_f401re_defconfig. Check 96b_carbon as a reference.

File: boards/arm/nucleo_f411re/nucleo_f411re.dts [hunk 4]
```diff
@@ -20,7 +20,7 @@
 
 &usart1 {
 	current-speed = <115200>;
-	pinctrl-0 = <&usart1_pins_a>;
+	pinctrl-0 = <&usart1_pins_b>;
```
Comment (ydamigos) [inline boards/arm/nucleo_f411re/nucleo_f411re.dts]: Why did you change the usart1 pins? Is there a conflict with I2C? If yes, you should also update board's doc and pinmux.c files.

Comment (erwango) [reply boards/arm/nucleo_f411re/nucleo_f411re.dts]: And ideally,you should split changes into 2 commits, one for I2C the other for uart and justify the change

Comment (rosterloh) [reply boards/arm/nucleo_f411re/nucleo_f411re.dts]: You are quite right. Thank must have been a mistake. Good catch.

File: boards/arm/nucleo_f411re/doc/nucleof411re.rst [hunk 5]
```diff
@@ -116,6 +117,10 @@ Default Zephyr Peripheral Mapping:
 - PWM_2_CH1 : PA0
 - USER_PB : PC13
 - LD2 : PA5
+- I2C1_SDA : PB9
+- I2C1_SCL : PB8
+- I2C2_SDA : PB2
```
Comment (ydamigos) [inline boards/arm/nucleo_f411re/doc/nucleof411re.rst]: PB3?

Comment (rosterloh) [reply boards/arm/nucleo_f411re/doc/nucleof411re.rst]: Done. Thought I might as well add I2C3 while I'm at it too

**Comment by ydamigos** [comment]
Shippable fails with:
`2: B4 Second line is not empty: "Add the required changes to support i2c on the Nucleo F411RE"`
Add an empty line between commit header and body.

**Comment by rosterloh** [comment]
@ydamigos I can't seem to amend my commit to have a newline at the start. Do you know how I can do this?

**Comment by ydamigos** [comment]
@rosterloh Use `git commit --amend`. After editing the commit message, push (using -f flag) it to your repository.

**Comment by rosterloh** [comment]
@ydamigos I don't know what was going on. That's exactly what I was doing but it seems to have passed now.

**Review by ydamigos — APPROVED** [review]
LGTM.

**Comment by galak** [comment]
@erwango and @superna9999 you guys good with this now?

**Review by erwango — APPROVED** [review]

**Comment by rosterloh** [comment]
@superna9999 I have done a rebase and it's down to you to approve.

**Review by galak — APPROVED** [review]
